sdk/dotnet/Pulumi/Config.cs

```diff
@@ -0,0 +1,196 @@
+            {
+                return v == null ? null : JsonDocument.Parse(v);
+            }
+            catch (JsonException ex)
+            {
+                throw new ConfigTypeException(FullKey(key), v, nameof(JsonDocument), ex);
+            }
+        }
+
+        /// <summary>
+        /// Loads an optional configuration value, as an object, by its key, marking it as a secret
+        /// or null if it doesn't exist. This routine simply JSON parses and doesn't validate the
+        /// shape of the contents.
+        /// </summary>
+        public Output<JsonDocument>? GetSecretJson(string key)
+            => MakeClassSecret(GetJson(key));
+
+        /// <summary>
+        /// Loads a configuration value by its given key.  If it doesn't exist, an error is thrown.
+        /// </summary>
+        public string Require(string key)
+            => Get(key) ?? throw new ConfigMissingException(FullKey(key));
+
+        /// <summary>
+        /// Loads a configuration value by its given key, marking it as a secet.  If it doesn't exist, an error
+        /// is thrown.
+        /// </summary>
+        public Output<string> RequireSecret(string key)
+            => MakeClassSecret(Require(key));
+
+        /// <summary>
+        /// Loads a configuration value, as a boolean, by its given key.  If it doesn't exist, or the
+        /// configuration value is not a legal boolean, an error is thrown.
+        /// </summary>
+        public bool RequireBoolean(string key)
+            => GetBoolean(key) ?? throw new ConfigMissingException(FullKey(key));
+
+        /// <summary>
+        /// Loads a configuration value, as a boolean, by its given key, marking it as a secret.
+        /// If it doesn't exist, or the configuration value is not a legal boolean, an error is thrown.
+        /// </summary>
+        public Output<bool> RequireSecretBoolean(string key)
+            => MakeStructSecret(RequireBoolean(key));
+
+        /// <summary>
+        /// Loads a configuration value, as a number, by its given key.  If it doesn't exist, or the
+        /// configuration value is not a legal number, an error is thrown.
+        /// </summary>
+        public int RequireInt32(string key)
+            => GetInt32(key) ?? throw new ConfigMissingException(FullKey(key));
+
+        /// <summary>
+        /// Loads a configuration value, as a number, by its given key, marking it as a secret.
+        /// If it doesn't exist, or the configuration value is not a legal number, an error is thrown.
+        /// </summary>
+        public Output<int> RequireSecretInt32(string key)
+            => MakeStructSecret(RequireInt32(key));
+
+        /// <summary>
+        /// Loads a configuration value as a JSON string and deserializes the JSON into a JavaScript
+        /// object. If it doesn't exist, or the configuration value is not a legal JSON string, an
+        /// error is thrown.
+        /// </summary>
+        public JsonDocument RequireJson(string key)
```

Comment:
Given that the value it is reading in is JSON, i'd prefer it be exposed as JSON.  Then we don't have to answer questions about things like: "what are lists stored as" and whatnot.

I would *highly* prefer to avoid dynamic unless there is a really good reason for it.

Reply:
I'd prefer keeping this closer in name to the API in [TypeScript](https://github.com/pulumi/pulumi/blob/06d4dde8898b2a0de2c3c7ff8e45f97495b89d82/sdk/nodejs/config.ts#L185)/[Python](https://github.com/pulumi/pulumi/blob/06d4dde8898b2a0de2c3c7ff8e45f97495b89d82/sdk/python/lib/pulumi/config.py#L178).

```typescript
public getObject<T>(key: string): T | undefined
```

```python
def get_object(self, key: str) -> Optional[Any]:
```

Why can't it be something like this (I have not tried it):

```c#
public T? GetObject<T>(string key)
{
    var v = Get(key);

    try
    {
        return v == null ? null : JsonSerializer.Deserialize<T>(v);
    }
    catch (JsonException ex)
    {
        throw new ConfigTypeException(FullKey(key), v, typeof(T).FullName, ex);
    }
}
```

That way, you can do things like:

```c#
var servers = config.GetObject<string[]>("servers");
```

...without having to manually parse the `JsonDocument`.

I'm also wary of having `JsonDocument` in our public API. Feels too close to `XmlDocument` for me (which fell out of favor and made APIs that included it in public signatures essentially legacy). Who knows when Microsoft will favor yet another API for parsing JSON in the future. If we don't tie ourselves to the types, we can easily swap out the implementation.

https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/system-xml-usage

> **X DO NOT** use `XmlNode` or `XmlDocument` to represent XML data. Favor using instances of `IXPathNavigable`, `XmlReader`, `XmlWriter`, or subtypes of `XNode` instead. `XmlNode` and `XmlDocument` are not designed for exposing in public APIs.

Reply:
sounds good.  will fixup.

Reply:
Another option is that we don't do anything here.  We just expose stuff as a string.  If it s really as easy as the user just calling JsonSerializer.Deserialize themselves, then they can do that.  That way they can also pass appropriate options to the Deserialize function and whatnot.

Reply:
have taken Justin's approach.